editor/src/clj/editor/code_completion.clj

```diff
@@ -13,55 +13,280 @@
+  (s/keys :req-un [::name ::display-string ::insert]
+          :opt-un [::type ::commit-characters ::detail ::doc]))
 
-(defn make-completions [resource-completion include-locals? namespace-alias]
-  (let [{:keys [vars local-vars functions local-functions]} resource-completion
-        var-info (if include-locals? (set/union vars local-vars) vars)
-        vars (map (fn [v] (lua/create-code-hint :variable v))
-                  var-info)
-        fn-info (if include-locals? (merge functions local-functions) functions)
-        fns  (map (fn [[fname {:keys [params]}]]
-                    (let [n (if namespace-alias (replace-alias fname namespace-alias) fname)
-                          display-string (str n "(" (string/join ", " params)  ")")]
-                      (lua/create-code-hint :function n display-string display-string "" {:select params})))
-                  fn-info)]
-    (set (concat vars fns))))
+(defn make
+  "Create a code completion
 
-(defn make-local-completions
-  [completion-info]
-  {"" (make-completions completion-info true nil)})
+  Required arg:
+    name    code completion identifier
 
-(defn make-required-completions
-  [requires required-completion-infos]
-  (let [required-completion-info-by-module (into {} (map (juxt :module identity)) required-completion-infos)]
-    (reduce (fn [ret [alias module]]
-              (let [completion-info (required-completion-info-by-module module)
-                    completions (make-completions completion-info false alias)
-                    module-name (or alias "")]
-                (merge-with set/union
-                            ret
-                            {module-name completions}
-                            (when alias {"" #{(lua/create-code-hint :namespace alias)}}))))
-            {}
-            requires)))
+  Optional kv-args:
+    :snippet              LSP-style snippet string for the code completion,
+                          mutually exclusive with :insert-string
+    :insert-string        literal snippet string for the code completion,
+                          mutually exclusive with :snippet
+    :display-string       code completion label shown in the completion popup
+    :type                 completion type, either Defold-specific :message or
+                          LSP-style :module, :function, :property etc.
+    :commit-characters    a set of one character long strings that accept the
+                          suggested completion if typed when the completion is
+                          selected
+    :detail               string with additional information to display in the
+                          completion popup, e.g. type or symbol information
+    :doc                  if a string, it is assumed to be a markdown
+                          documentation, otherwise, it can be either:
+                          - {:type :plaintext :value \"literal doc\"}
+                          - {:type :markdown  :value \"markdown doc\"}"
```

Comment:
I agree! I'll rename `:snippet` to `:insert-snippet` now, and will do the `:display-string` to `:label` later because I added some uses of `:display-string` during my current refactoring of the completion popup.